README.md
bubble_babble
=============

Bubble Babble encoding for nodejs.

Initially forked from tylorr/bubble\_babble to avoid the dependency on `Buffer`.

Notes:
The package available on NPMJS named `@keeex/bubble_babble` was based on this repository but got
more updates.
This repository is now archived and will not receive anymore updates.

**Install**
-------

The built package is not published on any repository; either install the
original package, or install from github:

```bash
npm install https://github.com/KeeeX/bubble_babble.git
```

**Usage**
-----

```JavaScript
import {encode, decode} from "@keeex/bubble_babble";

const encoded = encode("Pineapple");
console.log(encoded); // "xigak-nyryk-humil-bosek-sonax"

const ascii = decode("xesef-disof-gytuf-katof-movif-baxux");
console.log(ascii); // "1234567890"
```

Typescript
----------
This module includes type definition for usage with TypeScript


CHANGELOG.md
# Changelog

## v2.1.0 2023-06-22 12:19:11 +0200

### Added

- Preparation for publication
- Output of `encode()` can be changed from `ArrayBuffer` to either `Uint8Array` or an UTF-8 string

### Changed

- Moved to TypeScript

### Removed

- Manual types declaration

### Security

- Update

## v2.0.0 2021-06-04 17:35:31 +0200

### Changed

- Switched to "module" type of package

## v1.0.1 2021-06-04 17:32:12 +0200

### Security

- Update

## v1.0.0 2019-11-27 16:01:11 +0100

### Added

- Code coverage for tests
- Dev. dependencies for tests

### Changed

- [breaking] Package renamed to `@keeex/bubble_babble`
- [breaking] Uses import/export
- [breaking] Changed input/output type, removed dependency to Buffer
- [breaking] encode() only accept Uint8Array, ArrayBuffer, string
- [breaking] decode() returns ArrayBuffer

### Fixed

- Formatting
- Update

## v0.2.0

Forked from initial repository


test/test.js
/* eslint-disable no-magic-numbers */
import assert from "node:assert";
import {encode, decode} from "../bubble_babble.js";

const randomInt = (min, max) => Math.floor((Math.random() * (max - min + 1)) + min);

const LARGEST_BASE = 36;

const randomStr = () => {
  const randomSrc = Math.random().toString(LARGEST_BASE);
  return randomSrc.slice(2, randomInt(0, randomSrc.length));
};

const testVectors = [
  {
    ascii: "",
    encoding: "xexax",
  },
  {
    ascii: "1234567890",
    encoding: "xesef-disof-gytuf-katof-movif-baxux",
  },
  {
    ascii: "Pineapple",
    encoding: "xigak-nyryk-humil-bosek-sonax",
  },
];

const str2abUTF8 = str => new TextEncoder().encode(str).buffer;
const ab2strUTF8 = ab => new TextDecoder().decode(new Uint8Array(ab));

const testEncode = () => {
  it("should encode a buffer", () => testVectors.forEach(fixture => {
    const encoded = encode(str2abUTF8(fixture.ascii));
    assert.deepStrictEqual(encoded, fixture.encoding);
  }));

  it("should encode a string", () => testVectors.forEach(fixture => {
    const encoded = encode(fixture.ascii);
    assert.deepStrictEqual(encoded, fixture.encoding);
  }));

  it("should return a string that starts and begins with \"x\"", () => {
    for (let i = 0; i < 10; ++i) {
      const encoded = encode(randomStr());
      assert.ok(encoded.startsWith("x"));
      assert.ok(encoded.endsWith("x"));
    }
  });

  it("should be the inverse of decoding", () => testVectors.forEach(fixture => {
    const redecoded = encode(decode(fixture.encoding));
    assert.deepStrictEqual(redecoded, fixture.encoding);
  }));
};

const testDecode = () => {
  it("should decode a string and return a buffer", () => testVectors.forEach(fixture => {
    const decoded = decode(fixture.encoding);
    assert(decoded instanceof ArrayBuffer);
    const ascii = ab2strUTF8(decoded);
    assert.deepStrictEqual(ascii, fixture.ascii);
  }));

  it("should throw exception on corrupt input", () => {
    assert.throws(() => decode("xesyf-disof-gytuf-katof-movif-baxux"));
    assert.throws(() => decode("xesef-disof-gytuf-katof-movif-baxu"));
  });

  it("should be inverse of encoding a string", () => {
    const asciiInput = "Inverse of each other.";
    const reinput = ab2strUTF8(decode(encode(asciiInput)));
    assert.deepStrictEqual(reinput, asciiInput);
  });

  it("should accept Uint8Array", () => {
    const input = Uint8Array.from([34, 63, 42, 0, 255]);
    const decoded = decode(encode(input));
    assert.strictEqual(input.length, decoded.byteLength);
    const decodedBytes = new Uint8Array(decoded);
    for (let i = 0; i < decodedBytes.length; ++i) assert.strictEqual(decodedBytes[i], input[i]);
  });

  it("should be inverse of encoding an ArrayBuffer", () => {
    const input = Uint8Array.from([255, 255, 255, 255]).buffer;
    const decoded = decode(encode(input));
    assert.strictEqual(input.byteLength, decoded.byteLength);
    const decodedBytes = new Uint8Array(decoded);
    for (let i = 0; i < decodedBytes.length; ++i) assert.strictEqual(decodedBytes[i], 255);
  });
};

describe("BubbleBabble", () => {
  describe("#encode()", testEncode);
  describe("#decode()", testDecode);
});


src/bubble_babble.test.ts
/* eslint-disable no-magic-numbers */
import assert from "node:assert";
import {
  encode,
  decode,
} from "./bubble_babble.js";
import {OutputFormat} from "./types.js";

const randomInt = (
  min: number,
  max: number,
): number => Math.floor((Math.random() * (max - min + 1)) + min);

const LARGEST_BASE = 36;

const randomStr = (): string => {
  const randomSrc = Math.random().toString(LARGEST_BASE);
  return randomSrc.slice(2, randomInt(0, randomSrc.length));
};

const testVectors = [
  {
    ascii: "",
    encoding: "xexax",
  },
  {
    ascii: "1234567890",
    encoding: "xesef-disof-gytuf-katof-movif-baxux",
  },
  {
    ascii: "Pineapple",
    encoding: "xigak-nyryk-humil-bosek-sonax",
  },
];

const str2abUTF8 = (str: string): ArrayBuffer => new TextEncoder().encode(str).buffer;
const buf82strUTF8 = (buf8: Uint8Array): string => new TextDecoder().decode(buf8);
const ab2strUTF8 = (ab: ArrayBuffer): string => buf82strUTF8(new Uint8Array(ab));

const testEncode = () => {
  it("should encode a buffer", () => testVectors.forEach(fixture => {
    const encoded = encode(str2abUTF8(fixture.ascii));
    assert.deepStrictEqual(encoded, fixture.encoding);
  }));

  it("should encode a string", () => testVectors.forEach(fixture => {
    const encoded = encode(fixture.ascii);
    assert.deepStrictEqual(encoded, fixture.encoding);
  }));

  it("should return a string that starts and begins with \"x\"", () => {
    for (let i = 0; i < 10; ++i) {
      const encoded = encode(randomStr());
      assert.ok(encoded.startsWith("x"));
      assert.ok(encoded.endsWith("x"));
    }
  });

  it("should be the inverse of decoding", () => testVectors.forEach(fixture => {
    const redecoded = encode(decode(fixture.encoding));
    assert.deepStrictEqual(redecoded, fixture.encoding);
  }));
};

const testDecode = () => {
  it("should decode a string and return a buffer", () => testVectors.forEach(fixture => {
    const decoded = decode(fixture.encoding);
    assert(decoded instanceof ArrayBuffer);
    const ascii = ab2strUTF8(decoded);
    assert.deepStrictEqual(ascii, fixture.ascii);
  }));

  it("should return an Uint8Array after decode", () => testVectors.forEach(fixture => {
    const decoded = decode(fixture.encoding, OutputFormat.Uint8Array);
    assert(decoded instanceof Uint8Array);
    const ascii = buf82strUTF8(decoded);
    assert.deepStrictEqual(ascii, fixture.ascii);
  }));

  it("should return a string after decode", () => testVectors.forEach(fixture => {
    const decoded = decode(fixture.encoding, OutputFormat.string);
    assert(typeof decoded === "string");
    assert.deepStrictEqual(decoded, fixture.ascii);
  }));

  it("should throw exception on corrupt input", () => {
    assert.throws(() => decode("xesyf-disof-gytuf-katof-movif-baxux"));
    assert.throws(() => decode("xesef-disof-gytuf-katof-movif-baxu"));
    assert.throws(() => decode("xesef-disof-gytuf-katof-movif-bexux"));
  });

  it("should be inverse of encoding a string", () => {
    const asciiInput = "Inverse of each other.";
    const reinput = ab2strUTF8(decode(encode(asciiInput)));
    assert.deepStrictEqual(reinput, asciiInput);
  });

  it("should accept Uint8Array", () => {
    const input = Uint8Array.from([34, 63, 42, 0, 255]);
    const decoded = decode(encode(input));
    assert.strictEqual(input.length, decoded.byteLength);
    const decodedBytes = new Uint8Array(decoded);
    for (let i = 0; i < decodedBytes.length; ++i) assert.strictEqual(decodedBytes[i], input[i]);
  });

  it("should be inverse of encoding an ArrayBuffer", () => {
    const input = Uint8Array.from([255, 255, 255, 255]).buffer;
    const decoded = decode(encode(input));
    assert.strictEqual(input.byteLength, decoded.byteLength);
    const decodedBytes = new Uint8Array(decoded);
    for (let i = 0; i < decodedBytes.length; ++i) assert.strictEqual(decodedBytes[i], 255);
  });

};

describe("BubbleBabble", () => {
  describe("#encode()", testEncode);
  describe("#decode()", testDecode);
});


bubble_babble.js
const vowels = "aeiouy";
const vowelsCount = vowels.length;
const usedConsonants = "bcdfghklmnprstv";
const evenConsonant = "z";
const endsConsonant = "x";
const consonants = `${usedConsonants}${evenConsonant}${endsConsonant}`;
// Index of "x" in consonnants
const endIndex = 16;

/** Get an Uint8Array representation from input */
const getBuf8FromInput = input => {
  if (input instanceof Uint8Array) return input;
  if (input instanceof ArrayBuffer) return new Uint8Array(input);
  if (typeof input === "string") return new TextEncoder().encode(input);
  throw new Error("Unsupported input");
};

const oddPartial = (rawByte, checksum) => {
  /* eslint-disable no-magic-numbers */
  const a = (((rawByte >> 6) & 3) + checksum) % vowelsCount;
  const b = (rawByte >> 2) & 15;
  const c = ((rawByte & 3) + Math.floor(checksum / 6)) % vowelsCount;
  /* eslint-enable no-magic-numbers */
  return vowels[a] + consonants[b] + vowels[c];
};

const evenPartial = checksum => {
  /* eslint-disable no-magic-numbers */
  const a = checksum % vowelsCount;
  const c = Math.floor(checksum / 6);
  /* eslint-enable no-magic-numbers */
  return vowels[a] + endsConsonant + vowels[c];
};

/* eslint-disable no-magic-numbers */
const nextChecksum = (checksum, byte1, byte2) => ((checksum * 5) + (byte1 * 7) + byte2) % 36;
/* eslint-enable no-magic-numbers */

/** Encode a buffer/string into a bubble babble string */
export const encode = input => {
  const inputBytes = getBuf8FromInput(input);
  const len = inputBytes.length;
  let result = endsConsonant;
  let checksum = 1;
  let i = 0;
  // create full tuples
  while (i + 1 < len) {
    const byte1 = inputBytes[i++];
    const byte2 = inputBytes[i++];
    result += oddPartial(byte1, checksum);

    /* eslint-disable no-magic-numbers */
    const d = (byte2 >> 4) & 15;
    const e = byte2 & 15;
    /* eslint-enable no-magic-numbers */

    result = `${result}${consonants[d]}-${consonants[e]}`;
    checksum = nextChecksum(checksum, byte1, byte2);
  }
  // handle partial tuple
  if (i < len) {
    const byte1 = inputBytes[i];
    result += oddPartial(byte1, checksum);
  } else {
    result += evenPartial(checksum);
  }
  result += endsConsonant;
  return result;
};

const decodeTuple = asciiTuple => ([
  vowels.indexOf(asciiTuple[0]),
  consonants.indexOf(asciiTuple[1]),
  vowels.indexOf(asciiTuple[2]),
  consonants.indexOf(asciiTuple[3]),
  consonants.indexOf(asciiTuple[5]),
]);

const decode3PartByte = (a, b, c, checksum) => {
  /* eslint-disable no-magic-numbers */
  const high = (a - (checksum % 6) + 6) % 6;
  const mid = b;
  const low = (c - (Math.floor(checksum / 6) % 6) + 6) % 6;
  if (high >= 4 || low >= 4) throw new Error("Corrupt string");
  return (high << 6) | (mid << 2) | low;
  /* eslint-enable no-magic-numbers */
};

/* eslint-disable no-magic-numbers */
const decode2PartByte = (d, e) => (d << 4) | e;
/* eslint-enable no-magic-numbers */

export const decode = input => {
  if (input[0] !== endsConsonant || input.slice(-1) !== endsConsonant) {
    throw new Error("Corrupt string");
  }
  // eslint-disable-next-line require-unicode-regexp
  const asciiTuples = input.slice(1, -1).match(/.{3,6}/g);
  const len = asciiTuples ? asciiTuples.length : 0;
  const charCodes = [];
  let checksum = 1;
  let tuple;
  // handle full tuples
  for (let i = 0; i < len - 1; ++i) {
    tuple = decodeTuple(asciiTuples[i]);
    const byte1 = decode3PartByte(tuple[0], tuple[1], tuple[2], checksum);
    const byte2 = decode2PartByte(tuple[3], tuple[4]);
    checksum = nextChecksum(checksum, byte1, byte2);
    charCodes.push(byte1);
    charCodes.push(byte2);
  }

  // handle partial tuple
  tuple = decodeTuple(asciiTuples[len - 1]);
  if (tuple[1] === endIndex) {
    /* eslint-disable no-magic-numbers */
    if (tuple[0] !== checksum % 6 || tuple[2] !== Math.floor(checksum / 6)) {
      throw new Error("Corrupt string");
    }
    /* eslint-enable no-magic-numbers */
  } else {
    const byte1 = decode3PartByte(tuple[0], tuple[1], tuple[2], checksum);
    charCodes.push(byte1);
  }

  return Uint8Array.from(charCodes).buffer;
};


src/bubble_babble.ts
import {
  DecodedTuple,
  OutputFormat,
  SupportedInput,
} from "./types.js";

const vowels = "aeiouy".split("");
const vowelsCount = vowels.length;
const usedConsonants = "bcdfghklmnprstv".split("");
const evenConsonant = "z";
const endsConsonant = "x";
const consonants = [...usedConsonants, evenConsonant, endsConsonant];
// Index of "x" in consonnants
const endIndex = 16;
const separator = "-";

/** Get an Uint8Array representation from input */
const getBuf8FromInput = (input: SupportedInput): Uint8Array => {
  if (input instanceof Uint8Array) return input;
  if (input instanceof ArrayBuffer) return new Uint8Array(input);
  if (typeof input === "string") return new TextEncoder().encode(input);
  throw new Error("Unsupported input");
};

const oddPartial = (rawByte: number, checksum: number): Array<string> => {
  /* eslint-disable no-magic-numbers */
  const a = (((rawByte >> 6) & 3) + checksum) % vowelsCount;
  const b = (rawByte >> 2) & 15;
  const c = ((rawByte & 3) + Math.floor(checksum / 6)) % vowelsCount;
  /* eslint-enable no-magic-numbers */
  return [vowels[a], consonants[b], vowels[c]];
};

const evenPartial = (checksum: number): Array<string> => {
  /* eslint-disable no-magic-numbers */
  const a = checksum % vowelsCount;
  const c = Math.floor(checksum / 6);
  /* eslint-enable no-magic-numbers */
  return [vowels[a], endsConsonant, vowels[c]];
};

/* eslint-disable no-magic-numbers */
const nextChecksum = (
  checksum: number,
  byte1: number,
  byte2: number,
): number => ((checksum * 5) + (byte1 * 7) + byte2) % 36;
/* eslint-enable no-magic-numbers */

/** Encode a buffer/string into a bubble babble string */
export const encode = (input: SupportedInput): string => {
  const inputBytes = getBuf8FromInput(input);
  const len = inputBytes.length;
  let result = [endsConsonant];
  let checksum = 1;
  let i = 0;
  // create full tuples
  while (i + 1 < len) {
    const byte1 = inputBytes[i++];
    const byte2 = inputBytes[i++];
    result.push(...oddPartial(byte1, checksum));

    /* eslint-disable no-magic-numbers */
    const d = (byte2 >> 4) & 15;
    const e = byte2 & 15;
    /* eslint-enable no-magic-numbers */

    result.push(consonants[d], separator, consonants[e]);
    checksum = nextChecksum(checksum, byte1, byte2);
  }
  // handle partial tuple
  if (i < len) {
    const byte1 = inputBytes[i];
    result.push(...oddPartial(byte1, checksum));
  } else {
    result.push(...evenPartial(checksum));
  }
  result.push(endsConsonant);
  return result.join("");
};

const decodeTuple = (asciiTuple: string): DecodedTuple => ([
  vowels.indexOf(asciiTuple[0]),
  consonants.indexOf(asciiTuple[1]),
  vowels.indexOf(asciiTuple[2]),
  consonants.indexOf(asciiTuple[3]),
  consonants.indexOf(asciiTuple[5]),
]);

const decode3PartByte = (
  a: number,
  b: number,
  c: number,
  checksum: number,
): number => {
  /* eslint-disable no-magic-numbers */
  const high = (a - (checksum % 6) + 6) % 6;
  const mid = b;
  const low = (c - (Math.floor(checksum / 6) % 6) + 6) % 6;
  if (high >= 4 || low >= 4) throw new Error("Corrupt string");
  return (high << 6) | (mid << 2) | low;
  /* eslint-enable no-magic-numbers */
};

/* eslint-disable no-magic-numbers */
const decode2PartByte = (d: number, e: number): number => (d << 4) | e;
/* eslint-enable no-magic-numbers */

interface DecodeOverload {
  (input: string, output: OutputFormat.string): string;
  (input: string, output?: OutputFormat.ArrayBuffer): ArrayBuffer;
  (input: string, output: OutputFormat.Uint8Array): Uint8Array;
}

export const decode: DecodeOverload = (
  input: string,
  output: OutputFormat = OutputFormat.ArrayBuffer,
): any => {
  if (input[0] !== endsConsonant || input.slice(-1) !== endsConsonant) {
    throw new Error("Corrupt string");
  }
  // eslint-disable-next-line require-unicode-regexp
  const asciiTuples = input.slice(1, -1).match(/.{3,6}/g) as Array<string>;
  const len = asciiTuples ? asciiTuples.length : 0;
  const charCodes = [];
  let checksum = 1;
  let tuple;
  // handle full tuples
  for (let i = 0; i < len - 1; ++i) {
    tuple = decodeTuple(asciiTuples[i]);
    const byte1 = decode3PartByte(tuple[0], tuple[1], tuple[2], checksum);
    const byte2 = decode2PartByte(tuple[3], tuple[4]);
    checksum = nextChecksum(checksum, byte1, byte2);
    charCodes.push(byte1);
    charCodes.push(byte2);
  }

  // handle partial tuple
  tuple = decodeTuple(asciiTuples[len - 1]);
  if (tuple[1] === endIndex) {
    /* eslint-disable no-magic-numbers */
    if (tuple[0] !== checksum % 6 || tuple[2] !== Math.floor(checksum / 6)) {
      throw new Error("Corrupt string");
    }
    /* eslint-enable no-magic-numbers */
  } else {
    const byte1 = decode3PartByte(tuple[0], tuple[1], tuple[2], checksum);
    charCodes.push(byte1);
  }

  const buf8 = new Uint8Array(charCodes);
  if (output === OutputFormat.string) return new TextDecoder().decode(buf8);
  if (output === OutputFormat.ArrayBuffer) {
    if (buf8.byteOffset === 0 && buf8.byteLength === buf8.buffer.byteLength) return buf8.buffer;
    return buf8.buffer.slice(buf8.byteOffset, buf8.byteOffset + buf8.byteLength);
  }
  return buf8;
};
